Armadillo/src/main/java/com/scribd/armadillo/download/DownloadEngine.kt

```diff
@@ -39,39 +43,62 @@ internal interface DownloadEngine {
  * Starts the [DownloadService] when necessary
  */
 @Singleton
-internal class ExoplayerDownloadEngine @Inject constructor(private val context: Context,
-                                                           private val downloadHeadersStore: HeadersStore,
-                                                           private val downloadService: Class<out DownloadService>,
-                                                           private val downloadManager: DownloadManager,
-                                                           private val downloadTracker: DownloadTracker,
-                                                           private val stateModifier: StateStore.Modifier) : DownloadEngine {
+internal class ExoplayerDownloadEngine @Inject constructor(
+    private val context: Context,
+    private val downloadHeadersStore: HeadersStore,
+    private val downloadService: Class<out DownloadService>,
+    private val downloadManager: DownloadManager,
+    private val downloadTracker: DownloadTracker,
+    private val stateModifier: StateStore.Modifier,
+    private val offlineDrmManager: OfflineDrmManager,
+    @Named(Constants.DI.GLOBAL_SCOPE) private val globalScope: CoroutineScope,
+) : DownloadEngine {
     override fun init() = downloadTracker.init()
 
-    override fun download(audiobook: AudioPlayable) {
-        val downloadHelper = downloadHelper(context, audiobook.request)
+    override fun download(audioPlayable: AudioPlayable) {
+        globalScope.launch {
+            launch {
+                // Download DRM license for offline use
+                offlineDrmManager.downloadDrmLicenseForOffline(audioPlayable)
```

Comment:
This can throw an exception which would crash the app if there is DRM is included in an HLS stream. Should we add `CoroutineExceptionHandlers` or try/catches??

Reply:
Good point, I've added a coroutine exception handler to correctly dispatch the error to the client app without crashing it